openstack: Expose worker server group policy

pierreprinetti opened the pull request:
With this change, Compute nodes within each MachineSet are automatically
created in a Server group, with a default policy of
"soft-anti-affinity".

With this change, a "serverGroupPolicy" can be set in install-config, on
the worker MachinePool and/or in the platform default.

Implements [OSASINFRA-2570](https://issues.redhat.com/browse/OSASINFRA-2570)

Co-Authored-By: Matthew Booth <mbooth@redhat.com>

pierreprinetti commented:
/cc @SeanMooney

SeanMooney commented on docs/user/openstack/customization.md:
```diff
@@ -39,7 +39,7 @@ Beyond the [platform-agnostic `install-config.yaml` properties](../customization
 * `additionalNetworkIDs` (optional list of strings): IDs of additional networks for machines.
 * `additionalSecurityGroupIDs` (optional list of strings): IDs of additional security groups for machines.
-* `serverGroupPolicy` (optional string): Server group policy to apply to the group that will contain the machines in the pool. Defaults to "soft-anti-affinity". Only applicable to the Control plane MachinePool.
+* `serverGroupPolicy` (optional string): Server group policy to apply to the group that will contain the machines in the pool. Defaults to "soft-anti-affinity". Note that a strict "affinity" policy prevents migrations, and therefore affects OpenStack upgrades. Allowed values are "affinity", "soft-affinity", "anti-affinity", "soft-anti-affinity".
```
thanks for adding "Note that a strict "affinity" policy prevents migrations, and therefore affects OpenStack upgrades."

the other caveat that you may want to consider noting is that this cannot be changed after an instance is created so  if you want to change the member ship of server groups or the policy you will have to destroy the exstign workers and recretate the new ones with the new policy.

you can do that in a rolling fashion if you need too but the point being that the server group membership and policy for any worker instnace is fixed when its created for the lifetime of that worker.

pierreprinetti replied:
Added a note mentioning that a server group is for life. Thanks!

SeanMooney commented on pkg/tfvars/openstack/openstack.go:
```diff
@@ -174,14 +176,29 @@ func TFVars(masterConfigs []*v1alpha1.OpenstackProviderSpec, cloud string, exter
+	for _, workerConfig := range workerConfigs {
+		cfg.WorkerServerGroupNames = append(cfg.WorkerServerGroupNames, workerConfig.ServerGroupName)
+		if workerConfig.ServerGroupID != "" {
+			return nil, errors.Errorf("ServerGroupID is not implemented in the Installer. Please use ServerGroupName for automatic creation of the Compute server group.")
```
we generally recommend the use of the resouce UUIDs in external tools by the way.
names are intended to be human readable but uuids are preferred for external orchestrateors to uses.

you are allowed to have 2 server groups in the same tenant/project with the same name which is why the uuid is preferred.

pierreprinetti replied:
Very true. However I could not find a way to use UUID here.

The problem is that the Installer creates the Machine and MachineSet manifests (the declarative definition of the servers to be created) _before_ it hits the OpenStack API: the UUIDs are not available at that stage because the resources do not exist yet.

SeanMooney replied:
ah yes the server group create api does not allow you to choose the uuid
https://docs.openstack.org/api-ref/compute/?expanded=create-server-group-detail#create-server-group

we often allow the uuid to be specified to avoid this catch 22 situation but i guess this is not one of those cases.

mdbooth commented on pkg/asset/cluster/tfvars.go:
```diff
@@ -510,17 +510,38 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
+			}
+
+			for _, master := range masters {
+				masterSpecs = append(masterSpecs, master.Spec.ProviderSpec.Value.Object.(*openstackprovider.OpenstackProviderSpec))
```
In general, are we ok with a crash in the event that ProviderSpec is unset, or of an unexpected type?

pierreprinetti replied:
I believe that at this point, that property should be exactly of type `*openstackprovider.OpenstackProviderSpec`.

The alternative would be a controlled type assertion, but then what do we do in case it's not the type we expect? The answer heavily depends on how the user got to that state, which is information I don't know how to get.

mdbooth commented on pkg/asset/cluster/tfvars.go:
```diff
@@ -510,17 +510,38 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
 		}
+
+		var workermpool *openstack.MachinePool
+		if len(installConfig.Config.Compute) > 0 {
+			workermpool = installConfig.Config.Compute[0].Platform.OpenStack
+		}
```
This doesn't look right: we're only going to consider a server group defined in the first compute config.

pierreprinetti replied:
install-config's "Compute" slice only accepts one machinepool. I should add a comment.

pierreprinetti replied:
(added a comment)

patrickdillon commented:
Have you considered how changing the default behavior will affect existing workflows? Ideally, we don't want to break any workflows that previously worked. I know the RHV team ran into this for clusters running in the lab environment. Should also consider single-node clusters.

Just food for thought.

SeanMooney commented:
> Have you considered how changing the default behavior will affect existing workflows? Ideally, we don't want to break any workflows that previously worked. I know the RHV team ran into this for clusters running in the lab environment. Should also consider single-node clusters.
> 
> Just food for thought.

soft anti affinity is implemented as a weigher not a filter.
as such it never elimiates a host it just chagne the order in the weight list.

for single node deployment a list of one only has one order so the filter will have not impact other than doing a tiny be of extra work.

if you use the hard affinity/anti affinity polices (which we do not recommend) as a  default they are implemented as filters and it would be a problem but soft-anti-affinity should have no change in behavior for singel node.

this change should never result in a vm that could previously have booted form failing but it wil change the distibution of host to prefer spreading vms by default. the nova schedulers default behavior behaiovr is to spead anyway so you are just enhancing that perfernce with context of meber of the group.

so form an upgrade impact point of view this shoudl be minimal

if you uses any other policy as a default it would have a negitivy impact on exisitng deployments and workflows.
that is any policy beyond no policy.

pierreprinetti commented:
> Have you considered how changing the default behavior will affect existing workflows

Thank you for raising this. I've been discussing offline with @mdbooth and @SeanMooney about precisely this: if changing the default from "no-server-group" to "soft-anti-affinity" could have any negative impact.

One minor impact is the doubling of the Server group quota requirement; we are tracking related work in [OS.AS.INFRA-2571](https://issues.redhat.com/browse/OSASINFRA-2571).

That being said, in the 3-compute-node setup, the answer seems to be reassuring. "soft-anti-affinity" is a best-effort policy designed to be compatible with the most critical workflows (e.g. OSP upgrades), at the risk of being ineffective in some (not-so-corner) cases.

That is the rationale with which [we have already defaulted to soft-anti-affinity for masters in v4.5](https://issues.redhat.com/browse/OSASINFRA-1204).

Affinity policies should also have no impact on single-node deployments, as to the best of my knowledge, they never apply to the _first_ server in a group.

PS: I see that @SeanMooney has addressed the point much better than I ever could; thank you again for that.

pierreprinetti commented:
/cc mandre

pierreprinetti commented:
/retest

pierreprinetti commented:
/retest

pierreprinetti commented:
/retest

pierreprinetti commented:
/retest

pierreprinetti commented:
/hold
until we have QE ack for the feature

pierreprinetti commented:
@mdbooth @mandre 
I think I have addressed your comments. Can you please take a look and confirm that my responses and corrections are valid?

@patrickdillon 
Do you think that this change is acceptable in principle?

Even though we can't merge until QE acks for capacity, it would be optimal to have the approve/lgtm labels all lined up for merge, when the occasion presents itself.

pierreprinetti commented:
/retest

pierreprinetti commented:
/retest

pierreprinetti commented:
/test e2e-openstack-kuryr

pierreprinetti commented:
/retest

pierreprinetti commented:
One more comment (courtesy @itzikb-redhat ):

Documentation should also mention that for OpenStack migrations, you will need an extra free host for migrating from a host when you have `anti-affinity` on servers

pierreprinetti commented:
Added a note about migrating `anti-affinity` instances (thanks @itzikb-redhat)
/hold cancel

pierreprinetti commented:
/hold
While this patch is ready for review, we are still waiting for a QE ack.

pierreprinetti commented:
/test e2e-openstack-byon

pierreprinetti commented:
/assign @staebler

This is ready to merge once it gets approval and QE ack. The /hold is only in place to prevent accidental merging before QE confirms they have capacity to verify.

EmilienM commented on docs/user/openstack/README.md:
```diff
@@ -89,12 +89,14 @@ Once you configure the quota for your project, please ensure that the user for t
 The default deployment stands up 3 master nodes, which is the minimum amount required for a cluster. For each master node you stand up, you will need 1 instance, and 1 port available in your quota. They should be assigned a flavor with at least 16 GB RAM, 4 vCPUs, and 25 GB Disk (or Root Volume). It is theoretically possible to run with a smaller flavor, but be aware that if it takes too long to  ...
 
-The Master Nodes are placed in a single Server Group with "soft anti-affinity" policy; the machines will therefore be creted on separate hosts when possible.
+The Master Nodes are placed in a single Server Group with "soft anti-affinity" policy by default; the machines will therefore be creted on separate hosts when possible.
```
```suggestion
The Master Nodes are placed in a single Server Group with "soft anti-affinity" policy by default; the machines will therefore be created on separate hosts when possible.
```

EmilienM commented on docs/user/openstack/README.md:
```diff
@@ -89,12 +89,14 @@ Once you configure the quota for your project, please ensure that the user for t
 
 The default deployment stands up 3 worker nodes. Worker nodes host the applications you run on OpenShift. The flavor assigned to the worker nodes should have at least 2 vCPUs, 8 GB RAM and 25 GB Disk (or Root Volume). However, if you are experiencing `Out Of Memory` issues, or your installs are timing out, try increasing the size of your flavor to match the master nodes: 4 vCPUs and 16 GB RAM.
 
+The Master Nodes are placed in a single Server Group with "soft anti-affinity" policy by default; the machines will therefore be creted on separate hosts when possible.
```
```suggestion
The Master Nodes are placed in a single Server Group with "soft anti-affinity" policy by default; the machines will therefore be created on separate hosts when possible.
```

EmilienM commented on pkg/asset/machines/openstack/machines.go:
```diff
@@ -160,6 +157,7 @@ func generateProvider(clusterID string, platform *openstack.Platform, mpool *ope
 		PrimarySubnet:    platform.MachinesSubnet,
 		AvailabilityZone: az,
 		SecurityGroups:   securityGroups,
+		ServerGroupName:  clusterID + "-" + role + "-" + az,
```
are we sure we want `az` here? what if it's empty? do we end up with `cluster123-worker-` ?

pierreprinetti replied:
Leveraged the power of neural networks, PTAL

pierreprinetti replied:
it was ugly indeed.

EmilienM commented:
/lgtm

pierreprinetti commented:
The QE ack has landed.
/hold cancel

CC @staebler

pierreprinetti commented:
/retest

mandre commented on pkg/asset/cluster/tfvars.go:
```diff
@@ -510,17 +510,47 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
+			}
 		}
+
+		// Only considering the first Compute machinepool here, because
```
Let's try to not make assumptions that there will always only be one compute machinepool if possible, because that's going to bite us if the installer lifts the 1 worker machinepool restriction.

pierreprinetti replied:
Dropping this assumption will require us to accept a whole set of new assumptions.

At this stage, we have one MachineSet per Availability zone, and one machine-pool. As a consequence, we have multiple Server group names and one single Server group policy.

If we have multiple machine-pools, there is no straightforward way to match names with policies. The code changes required to support multiple machine-pools are non-negligeable, and I can't justify that effort since there is no proposal to support that case.

I have added an assertion that there is no more than one worker machine-pool. The code will crash loudly at the first test run, and we will clearly see the required change in that case.

WDYT?

mandre replied:
Fair enough. If it's too complex, it's not worth plan too much ahead and instead adding an obvious error message like you did is perfectly fine.

pierreprinetti commented:
/test e2e-aws

mandre commented on data/data/openstack/workers/variables.tf:
```diff
@@ -0,0 +1,9 @@
+variable "worker_server_group_names" {
```
We're already in the worker module, we don't need to prefix the variables with `worker`

mandre reviewed (commented):
Just a nit, then I think we're good to go.

pierreprinetti commented:
/test e2e-openstack

pierreprinetti commented:
/assign @staebler

mandre reviewed (commented):
/lgtm

pierreprinetti commented:
/retest

pierreprinetti commented:
/retest

mandre reviewed (commented):
Restoring my vote after the rebase.
/lgtm
/approve
